subprojects/solver/solver-smtlib/src/main/java/hu/bme/mit/theta/solver/smtlib/solver/SmtLibSolver.java

```diff
@@ -296,4 +335,101 @@ protected final GeneralResponse parseResponse(final String response) {
+            });
+            final var proof = termTransformer.toExpr(getModelResponse.getProof(), Bool(), new SmtLibModel(Collections.emptyMap()));
+            return proofFromExpr(proof);
+        } else {
+            throw new AssertionError();
+        }
+    }
+
+    private ProofNode proofFromExpr(Expr<BoolType> proof) {
+        checkState(proof instanceof FuncAppExpr<?, ?>, "Proof must be a function application.");
+        int id = 0;
+        final Map<Expr<?>, Integer> lookup = new LinkedHashMap<>();
+
+        final var args = extractFuncAndArgs((FuncAppExpr<?, ?>) proof).get2();
+
+        Deque<Expr<?>> proofStack = new LinkedList<>();
+        proofStack.push(args.get(0));
+        lookup.put(args.get(0), id++);
+
+        Expr<BoolType> root = cast(False(), Bool());
+        final var rootBuilder = new ProofNode.Builder(root);
+
+        Map<Integer, ProofNode.Builder> visited = new LinkedHashMap<>();
+        visited.put(lookup.get(args.get(0)), rootBuilder);
+
+        while (!proofStack.isEmpty()) {
+            final var proofNodeExpr = proofStack.pop();
+            if (!visited.containsKey(lookup.getOrDefault(proofNodeExpr, -1))) {
+                throw new Z3Exception("Node should exist in the graph nodes");
+            }
+            final var proofNode = visited.get(lookup.get(proofNodeExpr));
+
+            if (proofNodeExpr instanceof FuncAppExpr<?, ?> funcAppExpr) {
+                final var nameAndArgs = extractFuncAndArgs(funcAppExpr);
+                if (nameAndArgs.get1() instanceof RefExpr<?> refName && refName.getDecl().getName().startsWith("hyper-res")) {
+                    if (!nameAndArgs.get2().isEmpty()) {
+                        for (int i = 1; i < nameAndArgs.get2().size() - 1; ++i) {
+                            final var child = nameAndArgs.get2().get(i);
+                            if (!visited.containsKey(lookup.getOrDefault(child, -1))) {
+                                if (!lookup.containsKey(child)) {
+                                    lookup.put(child, id++);
+                                }
+                                visited.put(lookup.get(child), new Builder(extractProofExpr(child)));
+                                proofStack.push(child);
+                            }
+                            proofNode.addChild(visited.get(lookup.get(child)));
+                        }
+                    }
+                }
+            }
+        }
+        return rootBuilder.build();
+    }
+
+    private Expr<BoolType> extractProofExpr(Expr<?> expr) {
+        checkState(expr instanceof FuncAppExpr<?, ?>, "Proof should be function application.");
+        final var nameAndArgs = extractFuncAndArgs((FuncAppExpr<?, ?>) expr);
+        final var args = nameAndArgs.get2();
+        final var lastArg = args.get(args.size() - 1);
+        checkState(lastArg instanceof FuncAppExpr<?, ?>, "Proof should be function application.");
+        return (Expr<BoolType>) lastArg;
+    }
```

Comment:
Again. Maybe extracting these to a `HornSmtLibSolver extends SmtLibSolver` similarly how interpolation is not in this class? Not necessary, but I feel that `SmtLibSolver` is doing a lot of things.

Reply:
You're right, I'll extract these parts to a separate solver.